src/shelly_hap_temperature_sensor.cpp

```diff
@@ -0,0 +1,152 @@
+
+  mgos::ScopedCPtr name_owner((void *) cfg.name);
+  // Validation.
+  if (cfg.name != nullptr && strlen(cfg.name) > 64) {
+    return mgos::Errorf(STATUS_INVALID_ARGUMENT, "invalid %s",
+                        "name (too long, max 64)");
+  }
+  if (cfg.unit < 0 || cfg.unit > 1) {
+    return mgos::Errorf(STATUS_INVALID_ARGUMENT, "invalid unit");
+  }
+  // Now copy over.
+  if (cfg_->name != nullptr && strcmp(cfg_->name, cfg.name) != 0) {
+    mgos_conf_set_str(&cfg_->name, cfg.name);
+    *restart_required = true;
+  }
+  if (cfg_->unit != cfg.unit) {
+    cfg_->unit = cfg.unit;
+  }
+  return Status::OK();
+}
+
+Status TemperatureSensor::SetState(const std::string &state_json) {
+  (void) state_json;
+  return Status::OK();
+}
+
+void TemperatureSensor::ValueChanged() {
+  current_temperature_characteristic->RaiseEvent();
+}
+
+Status TemperatureSensor::Init() {
+  uint16_t iid = svc_.iid + 1;
+  current_temperature_characteristic = new mgos::hap::FloatCharacteristic(
+      iid++, &kHAPCharacteristicType_CurrentTemperature, 0.0, 100.0, 0.1,
+      std::bind(&TemperatureSensor::HandleCurrentTemperatureRead, this, _1, _2,
+                _3),
+      true /* supports_notification */, nullptr /* write_handler */,
+      kHAPCharacteristicDebugDescription_CurrentTemperature);
+  AddChar(current_temperature_characteristic);
+  AddChar(new mgos::hap::UInt8Characteristic(
+      iid++, &kHAPCharacteristicType_TemperatureDisplayUnits, 0, 255, 1,
+      std::bind(&TemperatureSensor::HandleTemperatureDisplayUnitsRead, this, _1,
+                _2, _3),
+      true /* supports_notification */,
+      std::bind(&TemperatureSensor::HandleTemperatureDisplayUnitsWrite, this,
+                _1, _2, _3),
+      kHAPCharacteristicDebugDescription_TemperatureDisplayUnits));
+  temp_sensor_->notifier_ = std::bind(&TemperatureSensor::ValueChanged, this);
+  LOG(LL_INFO, ("Exporting Temp"));
+
+  temp_sensor_->StartUpdating(5000);
+  return Status::OK();
+}
+
+HAPError TemperatureSensor::HandleCurrentTemperatureRead(
+    HAPAccessoryServerRef *server,
+    const HAPFloatCharacteristicReadRequest *request, float *value) {
+  float temp = static_cast<float>(temp_sensor_->GetTemperature().ValueOrDie());
+  *value = truncf(temp * 10) / 10;
+  (void) server;
+  (void) request;
+  return kHAPError_None;
+}
+
+HAPError TemperatureSensor::HandleTemperatureDisplayUnitsWrite(
+    HAPAccessoryServerRef *server,
+    const HAPUInt8CharacteristicWriteRequest *request, uint8_t value) {
+  (void) value;
+  (void) server;
+  (void) request;
+  return kHAPError_None;
+}
+
+HAPError TemperatureSensor::HandleTemperatureDisplayUnitsRead(
+    HAPAccessoryServerRef *server,
+    const HAPUInt8CharacteristicReadRequest *request, uint8_t *value) {
+  *value = cfg_->unit;
+  (void) server;
+  (void) request;
+  return kHAPError_None;
+}
+
+StatusOr<std::string> TemperatureSensor::GetInfo() const {
+  return mgos::SPrintf("v: %f", temp_sensor_->GetTemperature().ValueOrDie());
```

Comment:
never use ValueOrDie without checking if it's ok first. we don't want firmware to die, ever.